File: static/js/smart_filter.js
/**
 * static/js/smart_filter.js — Filtro Avançado (Smart Filter)
 *
 * Utilitário reutilizável para filtros com sintaxe avançada.
 *
 * Tipos suportados:
 *   Número : "1,3,5-10,11"          → lista de inteiros expandida
 *   Texto  : '"txt1","*parte*"'     → lista de termos com curingas
 *   Select : <select multiple>      → usar getMultiSelectValues()
 */

/**
 * Parseia filtro numérico: "1,3,5-10,11" → [1, 3, 5, 6, 7, 8, 9, 10, 11]
 * @param {string} str
 * @returns {number[]}
 */
export function parseSmartNumber(str) {
  if (!str || !str.trim()) return [];
  const result = new Set();
  for (const part of str.split(',')) {
    const t = part.trim();
    if (!t) continue;
    const rangeMatch = t.match(/^(\d+)-(\d+)$/);
    if (rangeMatch) {
      const a = parseInt(rangeMatch[1], 10);
      const b = parseInt(rangeMatch[2], 10);
      if (a <= b && (b - a) <= 10_000) {
        for (let i = a; i <= b; i++) result.add(i);
      }
    } else {
      const n = parseInt(t, 10);
      if (!isNaN(n)) result.add(n);
    }
  }
  return [...result].sort((a, b) => a - b);
}

/**
 * Parseia filtro de texto com curingas.
 * Separa por vírgula; aspas ao redor do termo são opcionais e ignoradas.
 * * no início/fim indica curinga.
 * Ex: REF001,*LEROY*,COMEÇA*,*TERMINA
 * Ex (com aspas, legado): "REF001","*LEROY*"
 * @param {string} str
 * @returns {{ type: 'exact'|'contains'|'startswith'|'endswith', value: string }[]}
 */
export function parseSmartText(str) {
  if (!str || !str.trim()) return [];
  const terms = [];
  for (const part of str.split(',')) {
    // Remove aspas opcionais ao redor do termo (suporte legado)
    const raw = part.trim().replace(/^"|"$/g, '');
    if (!raw) continue;
    const start = raw.startsWith('*');
    const end   = raw.endsWith('*');
    if (start && end) {
      terms.push({ type: 'contains',   value: raw.slice(1, -1) });
    } else if (start) {
      terms.push({ type: 'endswith',   value: raw.slice(1) });
    } else if (end) {
      terms.push({ type: 'startswith', value: raw.slice(0, -1) });
    } else {
      terms.push({ type: 'exact',      value: raw });
    }
  }
  return terms;
}

/**
 * Retorna os valores selecionados de um <select multiple>.
 * @param {HTMLSelectElement} selectEl
 * @returns {string[]}
 */
export function getMultiSelectValues(selectEl) {
  return Array.from(selectEl.selectedOptions).map(o => o.value).filter(Boolean);
}

/**
 * Valida a sintaxe de um filtro numérico.
 * @param {string} str
 * @returns {boolean}
 */
export function validateSmartNumber(str) {
  if (!str || !str.trim()) return true;
  return /^[\d\s,\-]+$/.test(str.trim());
}

/**
 * Valida a sintaxe de um filtro de texto.
 * Aceita qualquer string não vazia (aspas são opcionais).
 * @param {string} str
 * @returns {boolean}
 */
export function validateSmartText(str) {
  if (!str || !str.trim()) return true;
  // Termos separados por vírgula; basta ter ao menos um caractere não vazio
  return str.split(',').some(p => p.trim().replace(/^"|"$/g, '').length > 0);
}

File: sac_base/smart_filter.py
"""
sac_base/smart_filter.py — Filtro Avançado (Smart Filter) — Backend

Utilitário reutilizável que espelha a lógica de static/js/smart_filter.js.

Tipos:
  Número : "1,3,5-10,11"          → lista de inteiros expandida
  Texto  : '"txt1","*parte*"'     → lista de dicts com tipo e valor
  Select : lista de strings já parseada pelo frontend
"""

import re
from django.db.models import Q


def parse_smart_number(value: str) -> list:
    """
    Parseia filtro numérico: "1,3,5-10,11" → [1, 3, 5, 6, 7, 8, 9, 10, 11]
    Limita expansão de intervalos a 10.000 elementos por segurança.
    """
    if not value or not value.strip():
        return []
    result: set = set()
    for part in value.split(','):
        t = part.strip()
        if not t:
            continue
        m = re.match(r'^(\d+)-(\d+)$', t)
        if m:
            a, b = int(m.group(1)), int(m.group(2))
            if a <= b and (b - a) <= 10_000:
                result.update(range(a, b + 1))
        else:
            try:
                result.add(int(t))
            except ValueError:
                pass
    return sorted(result)


def parse_smart_text(value: str) -> list:
    """
    Parseia filtro de texto: 'REF001,*LEROY*' ou (legado) '"REF001","*LEROY*"' ->
    [{'type': 'exact', 'value': 'REF001'}, {'type': 'contains', 'value': 'LEROY'}]
    Separa por virgula; aspas ao redor do termo sao opcionais e ignoradas.
    """
    if not value or not value.strip():
        return []
    terms = []
    for part in value.split(','):
        # Remove aspas opcionais ao redor do termo (suporte legado)
        raw = re.sub(r'^"(.*)"$', r'\1', part.strip())
        if not raw:
            continue
        start = raw.startswith('*')
        end   = raw.endswith('*')
        if start and end:
            terms.append({'type': 'contains',   'value': raw[1:-1]})
        elif start:
            terms.append({'type': 'endswith',   'value': raw[1:]})
        elif end:
            terms.append({'type': 'startswith', 'value': raw[:-1]})
        else:
            terms.append({'type': 'exact',      'value': raw})
    return terms


def apply_smart_number_filter(qs, field: str, value: str):
    """Aplica filtro numérico ao queryset. Sem valor, retorna qs sem alteração."""
    ids = parse_smart_number(value)
    if ids:
        qs = qs.filter(**{f'{field}__in': ids})
    return qs


def apply_smart_text_filter(qs, field: str, value: str):
    """
    Aplica filtro de texto ao queryset.
    Os termos são combinados com OR; sem valor, retorna qs sem alteração.
    """
    terms = parse_smart_text(value)
    if not terms:
        return qs
    q = Q()
    for term in terms:
        t, v = term['type'], term['value']
        if t == 'exact':
            q |= Q(**{field: v})
        elif t == 'contains':
            q |= Q(**{f'{field}__icontains': v})
        elif t == 'startswith':
            q |= Q(**{f'{field}__istartswith': v})
        elif t == 'endswith':
            q |= Q(**{f'{field}__iendswith': v})
    return qs.filter(q)
